readme.md

```diff
@@ -15,834 +14,1113 @@
+	var ret;
 
-    B. Faces of `this`
+	if ( foo ) {
+		ret = 'foo';
+	} else {
+		ret = 'quux';
+	}
+	return ret;
+}
 
-    Keep `this` "local" to the class. This means &ndash; do not bind e.g. a function in `selection.js` file to editor instance and do not create a function which should be ran in editor context. This way you'll avoid confusion what `this` refers to.
+// Good:
+function returnEarly( foo ) {
+	if ( foo ) {
+		return 'foo';
+	}
+	return 'quux';
+}
 
-    ```javascript
+// Bad:
+function doSthComplex( foo, bar ) {
+	if ( foo ) {
+		// do something
 
-    // 6.B.1
-    function Selection() {
-        var editor = this.editor;
+		if ( bar ) {
+			// do something
+		} else {
+			// do something else
+		}
+	}
+}
 
-        // Bad:
-        editor.on( 'setData', someCallback, editor );
+// Good:
+function doSthComplex( foo, bar ) {
+	if ( !foo )
+		return;
 
-        // Still bad:
-        editor.on( 'setData', CKEDITOR.tools.bind( someCallback, editor ) );
+	// do something
 
-        // Good:
-        editor.on( 'setData', someCallback, this );
+	if ( !bar ) {
+		// do something else
+		return;
+	}
 
-        // Good:
-        this.getEditor().on( 'setData', function( evt ) {
-            evt.editor.doSomething();
-        } );
+	// do something
+}
+```
 
-        // Good too:
-        var that = this;
-        editor.on( 'setData', function( evt ) {
-            editor.setCurrentSelection( that );
-        } );
-    }
+### 2.8.2. Objects minify poorly, closures very well
 
-    function someCallback() {
-        // Should refers to Selection, not Editor.
-        this;
-    }
+Avoid excessing usage of objects, when functional programming may be used as well. Properties and method names unlike private variables and function names cannot be shortened by minifier.
+You can play with [online Google Closure Compiler](http://closure-compiler.appspot.com/home) to test different approaches. Make sure to check the gzipped size which matters the most.
 
-    ```
 
-    C. Use `thisArg`
+## 2.9. Native & Host Objects
 
-    Several prototype methods of ES 5.1 built-ins come with a special `thisArg` signature, which should be used whenever possible. CKEditor follows this practice too, so for example [`eventTarget.on`](http://docs.ckeditor.com/#!/api/CKEDITOR.event-method-on) also accepts context as an argument.
+The basic principle here is:
 
-    ```javascript
+	Don't do stupid shit and everything will be ok.
 
-    // 6.C.1
+To reinforce this concept, please watch the following presentation:
+	
+[“Everything is Permitted: Extending Built-ins” by Andrew Dupont (JSConf2011, Portland, Oregon)](https://www.youtube.com/watch?v=xL3xCO7CLNM)
 
-    var obj = { f: 'foo', b: 'bar', q: 'qux' };
+**But** CKEditor is a widget - an application inside someone's system &ndash; so... do not touch anything except
+`CKEDITOR.*` namespace.
 
-    Object.keys( obj ).forEach( function( key ) {
 
-      // |this| now refers to `obj`
+## 2.10. Comments
```

Comment:
I'm not sure where to put this. Maybe better would be the section about conditions. However, there should be explicit written that comment to a given if should be before it not inside it.

Reply:
Please add info about ticket referencing convention.